### PR TITLE
fix: parse opus file extensions correctly

### PR DESCRIPTION
**Pull request by du-cki**
# Parse "audio/ogg; codecs=opus" correctly.

## Context

This brief PR fixes the `determineMediaExtension` method, because previously it was ambiguously parsing a file extension from its content type by just splitting by `/` and returning the last element (with other exceptions), which is fine for most use-cases but when a content type like `audio/ogg; codecs=opus` passes through, the file's extension is said to be `.ogg; codecs=opus` as 
evident by the logs: 

```sh
time="<time>" level=info msg="Chatwoot: Found audio attachment at statics/media/<uuid>.ogg; codecs=opus"
```

The fix I've implemented just strips off the `; codecs=opus` (or more specifically, anything after `;`, if it exists) and lets the system MIME DB handle the `audio/ogg` (which gets resolved as `.oga`).

Hopefully this PR also fixes audio files being sent to Chatwoot from GOWA, I've yet to test it in Chatwoot as I lack a testing environment as of right now, but from what I understand it should work as intended. I'm happy to make any changes if need be though.
image

## Test Results

I've added a new test-case for opus audio in the unit tests, and have manually tested it along with the entire application and its all passing.
<details>
<summary>Unit Test Logs</summary>

```sh
=== RUN   TestDetermineMediaExtension
=== RUN   TestDetermineMediaExtension/DocxFromFilename
=== RUN   TestDetermineMediaExtension/XlsxFromMime
=== RUN   TestDetermineMediaExtension/PptxFromMime
=== RUN   TestDetermineMediaExtension/ZipFallback
=== RUN   TestDetermineMediaExtension/AudioOgaWithCodecsParam
=== RUN   TestDetermineMediaExtension/ExeFromFilename
--- PASS: TestDetermineMediaExtension (0.00s)
    --- PASS: TestDetermineMediaExtension/DocxFromFilename (0.00s)
    --- PASS: TestDetermineMediaExtension/XlsxFromMime (0.00s)
    --- PASS: TestDetermineMediaExtension/PptxFromMime (0.00s)
    --- PASS: TestDetermineMediaExtension/ZipFallback (0.00s)
    --- PASS: TestDetermineMediaExtension/AudioOgaWithCodecsParam (0.00s)
    --- PASS: TestDetermineMediaExtension/ExeFromFilename (0.00s)
=== RUN   TestEnvironmentTestSuite
=== RUN   TestEnvironmentTestSuite/TestEnv
=== RUN   TestEnvironmentTestSuite/TestIsLocal
=== RUN   TestEnvironmentTestSuite/TestIsLocal/Production_environment
=== RUN   TestEnvironmentTestSuite/TestIsLocal/Staging_environment
=== RUN   TestEnvironmentTestSuite/TestIsLocal/Integration_environment
=== RUN   TestEnvironmentTestSuite/TestIsLocal/Development_environment
=== RUN   TestEnvironmentTestSuite/TestIsLocal/Local_environment
=== RUN   TestEnvironmentTestSuite/TestLoadConfig
=== RUN   TestEnvironmentTestSuite/TestMustHaveEnv
time="2026-03-08T20:21:05+04:00" level=warning msg="context.Background map[field:ENV_FROM_FILE]variable is not well set, reading from .env file"
=== RUN   TestEnvironmentTestSuite/TestMustHaveEnvBool
=== RUN   TestEnvironmentTestSuite/TestMustHaveEnvInt
=== RUN   TestEnvironmentTestSuite/TestMustHaveEnvMinuteDuration
--- PASS: TestEnvironmentTestSuite (0.00s)
    --- PASS: TestEnvironmentTestSuite/TestEnv (0.00s)
    --- PASS: TestEnvironmentTestSuite/TestIsLocal (0.00s)
        --- PASS: TestEnvironmentTestSuite/TestIsLocal/Production_environment (0.00s)
        --- PASS: TestEnvironmentTestSuite/TestIsLocal/Staging_environment (0.00s)
        --- PASS: TestEnvironmentTestSuite/TestIsLocal/Integration_environment (0.00s)
        --- PASS: TestEnvironmentTestSuite/TestIsLocal/Development_environment (0.00s)
        --- PASS: TestEnvironmentTestSuite/TestIsLocal/Local_environment (0.00s)
    --- PASS: TestEnvironmentTestSuite/TestLoadConfig (0.00s)
    --- PASS: TestEnvironmentTestSuite/TestMustHaveEnv (0.00s)
    --- PASS: TestEnvironmentTestSuite/TestMustHaveEnvBool (0.00s)
    --- PASS: TestEnvironmentTestSuite/TestMustHaveEnvInt (0.00s)
    --- PASS: TestEnvironmentTestSuite/TestMustHaveEnvMinuteDuration (0.00s)
=== RUN   TestUtilsTestSuite
=== RUN   TestUtilsTestSuite/TestContainsMention
=== RUN   TestUtilsTestSuite/TestContainsMention/should_success_get_phone_when_@_with_space
=== RUN   TestUtilsTestSuite/TestContainsMention/should_success_get_phone_without_suffix_space
=== RUN   TestUtilsTestSuite/TestContainsMention/should_success_get_phone_without_prefix_space
=== RUN   TestUtilsTestSuite/TestCreateFolder
=== RUN   TestUtilsTestSuite/TestCreateFolderEdgeCases
=== RUN   TestUtilsTestSuite/TestDownloadAudioFromURL
time="2026-03-08T20:21:05+04:00" level=warning msg="DownloadAudioFromURL: unexpected content type 'text/html', proceeding anyway"
=== RUN   TestUtilsTestSuite/TestDownloadImageFromURL
=== RUN   TestUtilsTestSuite/TestDownloadImageFromURLEdgeCases
=== RUN   TestUtilsTestSuite/TestDownloadVideoFromURL
=== RUN   TestUtilsTestSuite/TestGetMetaDataFromURL
time="2026-03-08T20:21:05+04:00" level=warning msg="Image download failed with status: 404 Not Found"
=== RUN   TestUtilsTestSuite/TestGetMetaDataFromURLEdgeCases
time="2026-03-08T20:21:05+04:00" level=warning msg="URL returned non-image content type: text/html; charset=utf-8"
time="2026-03-08T20:21:20+04:00" level=warning msg="Failed to decode image: image: unknown format"
=== RUN   TestUtilsTestSuite/TestGetMetaDataFromURLSendsBrowserHeaders
time="2026-03-08T20:21:25+04:00" level=warning msg="Failed to decode image: image: unknown format"
=== RUN   TestUtilsTestSuite/TestPanicIfNeeded
=== RUN   TestUtilsTestSuite/TestPanicIfNeededEdgeCases
=== RUN   TestUtilsTestSuite/TestRemoveFile
=== RUN   TestUtilsTestSuite/TestRemoveFileEdgeCases
=== RUN   TestUtilsTestSuite/TestStrToFloat64
=== RUN   TestUtilsTestSuite/TestStrToFloat64EdgeCases
--- PASS: TestUtilsTestSuite (21.11s)
    --- PASS: TestUtilsTestSuite/TestContainsMention (0.00s)
        --- PASS: TestUtilsTestSuite/TestContainsMention/should_success_get_phone_when_@_with_space (0.00s)
        --- PASS: TestUtilsTestSuite/TestContainsMention/should_success_get_phone_without_suffix_space (0.00s)
        --- PASS: TestUtilsTestSuite/TestContainsMention/should_success_get_phone_without_prefix_space (0.00s)
    --- PASS: TestUtilsTestSuite/TestCreateFolder (0.00s)
    --- PASS: TestUtilsTestSuite/TestCreateFolderEdgeCases (0.00s)
    --- PASS: TestUtilsTestSuite/TestDownloadAudioFromURL (0.01s)
    --- PASS: TestUtilsTestSuite/TestDownloadImageFromURL (0.00s)
    --- PASS: TestUtilsTestSuite/TestDownloadImageFromURLEdgeCases (0.00s)
    --- PASS: TestUtilsTestSuite/TestDownloadVideoFromURL (0.00s)
    --- PASS: TestUtilsTestSuite/TestGetMetaDataFromURL (0.06s)
    --- PASS: TestUtilsTestSuite/TestGetMetaDataFromURLEdgeCases (20.01s)
    --- PASS: TestUtilsTestSuite/TestGetMetaDataFromURLSendsBrowserHeaders (0.00s)
    --- PASS: TestUtilsTestSuite/TestPanicIfNeeded (0.00s)
    --- PASS: TestUtilsTestSuite/TestPanicIfNeededEdgeCases (0.00s)
    --- PASS: TestUtilsTestSuite/TestRemoveFile (0.00s)
    --- PASS: TestUtilsTestSuite/TestRemoveFileEdgeCases (1.00s)
    --- PASS: TestUtilsTestSuite/TestStrToFloat64 (0.00s)
    --- PASS: TestUtilsTestSuite/TestStrToFloat64EdgeCases (0.00s)
=== RUN   TestResponseTestSuite
=== RUN   TestResponseTestSuite/TestResponseDataJSON
=== RUN   TestResponseTestSuite/TestResponseDataJSONUnmarshaling
=== RUN   TestResponseTestSuite/TestResponseDataStruct
=== RUN   TestResponseTestSuite/TestResponseDataWithEmptyResults
=== RUN   TestResponseTestSuite/TestResponseDataWithNilResults
=== RUN   TestResponseTestSuite/TestResponseDataZeroValues
--- PASS: TestResponseTestSuite (0.00s)
    --- PASS: TestResponseTestSuite/TestResponseDataJSON (0.00s)
    --- PASS: TestResponseTestSuite/TestResponseDataJSONUnmarshaling (0.00s)
    --- PASS: TestResponseTestSuite/TestResponseDataStruct (0.00s)
    --- PASS: TestResponseTestSuite/TestResponseDataWithEmptyResults (0.00s)
    --- PASS: TestResponseTestSuite/TestResponseDataWithNilResults (0.00s)
    --- PASS: TestResponseTestSuite/TestResponseDataZeroValues (0.00s)
PASS
ok      github.com/aldinokemal/go-whatsapp-web-multidevice/pkg/utils    21.129s
```
</details>

### CODE REVIEW

File: src/infrastructure/chatwoot/client.go [hunk 1]
```diff
@@ -516,10 +516,15 @@ func (c *Client) createMessageWithAttachments(endpoint, content, messageType str
...
 
-			mimeType := mime.TypeByExtension(filepath.Ext(fp))
+			mimeType := mime.TypeByExtension(ext)
 			if mimeType == "" {
-				mimeType = "application/octet-stream"
+				if ext == ".oga" {
+					mimeType = "audio/ogg"
+				} else {
+					mimeType = "application/octet-stream"
+				}
```
[the remark this answers is not in the document]
Comment (du-cki) [reply src/infrastructure/chatwoot/client.go:527]: I'm sure this specific edge-case is not a real problem in production as the filepath is not user provided and is only called internally. I've also not made any major changes, and just added one more if-case here. The previous code had no such thing.

File: src/infrastructure/chatwoot/sync.go [hunk 2]
```diff
@@ -368,7 +368,7 @@ func getExtensionForMediaType(mediaType, filename string) string {
 	case "video":
 		return ".mp4"
 	case "audio", "ptt":
-		return ".ogg"
+		return ".oga"
```
[the remark this answers is not in the document]
Comment (du-cki) [reply src/infrastructure/chatwoot/sync.go]: For backwards compatibility sake, should I just default it to `.ogg` then, rather than the `.oga`? I don't think it makes much of a difference, considering when its in practice from what I understand.

File: src/infrastructure/chatwoot/client.go [hunk 3]
```diff
@@ -516,10 +516,15 @@ func (c *Client) createMessageWithAttachments(endpoint, content, messageType str
 			defer file.Close()
 
 			fileName := filepath.Base(fp)
+			ext := filepath.Ext(fp)
```
Comment (aldinokemal) [inline src/infrastructure/chatwoot/client.go:519]: ### Case-Sensitivity Issue

`filepath.Ext()` preserves the original case of the file extension. While in practice our codebase generates lowercase extensions, this is fragile if a file ever has an uppercase extension (e.g., from a user-provided filename).

Lowercasing `ext` here makes both the `mime.TypeByExtension()` call and the `.oga` comparison below case-insensitive. (`mime.TypeByExtension` already lowercases internally, so this is mainly for the `.oga` check.)

```suggestion
						ext := strings.ToLower(filepath.Ext(fp))
```

`strings` is already imported in this file ✅

File: src/infrastructure/chatwoot/client.go [hunk 4]
```diff
@@ -516,10 +516,15 @@ func (c *Client) createMessageWithAttachments(endpoint, content, messageType str
...
 
-			mimeType := mime.TypeByExtension(filepath.Ext(fp))
+			mimeType := mime.TypeByExtension(ext)
 			if mimeType == "" {
-				mimeType = "application/octet-stream"
+				if ext == ".oga" {
```
Comment (aldinokemal) [inline src/infrastructure/chatwoot/client.go:523]: ### Why This Fallback Is Needed

For context: after the root fix in `whatsapp.go`, `determineMediaExtension("", "audio/ogg; codecs=opus")` will strip to `audio/ogg`, and `mime.ExtensionsByType("audio/ogg")` returns `.oga` on most platforms.

So files saved by `ExtractMedia()` WILL have `.oga` extension. But Go's `mime.TypeByExtension(".oga")` returns `""` on some platforms (notably macOS) because `.oga` isn't in the system MIME database. This fallback correctly handles that gap.

**This is the right fix** — just needs the case-insensitive `ext` from the suggestion above. 👍

File: src/pkg/utils/whatsapp.go [hunk 5]
```diff
@@ -73,6 +73,10 @@ func determineMediaExtension(originalFilename, mimeType string) string {
 		}
 	}
 
+	if idx := strings.Index(mimeType, ";"); idx >= 0 {
```
Comment (aldinokemal) [inline src/pkg/utils/whatsapp.go:76]: ### Well-Placed Fix ✅

Good placement — stripping MIME parameters here (before `resolveKnownDocumentExtension` and `mime.ExtensionsByType`) ensures all downstream lookups receive a clean MIME type.

This also protects against theoretical edge cases like `application/vnd.openxmlformats-officedocument.wordprocessingml.document; charset=utf-8` hitting the known-document map — though I'd suggest adding a test case for that too (see overall review).

File: src/infrastructure/chatwoot/client.go (hunk 3 above)
Comment (du-cki) [reply src/infrastructure/chatwoot/client.go:519]: Like I've disputed with the AI Agent before, this is usually not an actual use-case and this is irrelevant to my PR anyway. As far as I looked, the filepaths seemed to be something that's passed around within internal code.

**Comment by HERRKIN** [comment]
I can confirm this issue on my side as well.

Current behavior:
- Chatwoot receives incoming audio from GOWA as a downloadable attachment instead of an inline playable audio message.
- This happens with WhatsApp voice notes and also with regular MP3 audio files.
- The same Chatwoot instance works correctly with WAHA, so this appears specific to GOWA -> Chatwoot media handling.

What I observed:
- Final Active Storage response returns:
  - Content-Type: audio/opus
  - Content-Disposition: attachment
- Example filename ends up malformed, including codec information:
  - 1773880769-f54b07fb-a334-4cd9-9dbe-964fa17c9256.ogg- codecs=opus

Headers from final file response:
- Content-Type: audio/opus
- Content-Disposition: attachment; filename="1773880769-f54b07fb-a334-4cd9-9dbe-964fa17c9256.ogg- codecs%3Dopus"

Notes:
- This PR looks very relevant for the malformed filename / codec suffix problem.
- However, in my case, even regular MP3 audio is still shown as a downloadable attachment in Chatwoot rather than an inline playable audio message.
- So this PR may fix part of the problem, especially for opus voice notes, but there may still be another issue affecting audio attachments more generally.

**Comment by HERRKIN** [comment]
I tested this on the latest version and the issue is still present.
The audio does not appear as a playable voice/audio message. It is shown as a regular downloadable file instead.

So the problem does not seem fully fixed yet.

**Comment by du-cki** [comment]
Ah yeah, I've suspected that much but I just didn't have the time to dive into this problem. I hoped fixing the immediate problem would at least help. If I do get time around the weekends or somewhere later down the line, I can look into it thoroughly going through Chatwoot's API documentation. If you do end up finding a solution for the problem, please mention me in the PR so I can watch its status.
